packages/flutter/lib/src/widgets/framework.dart

```diff
@@ -879,12 +891,28 @@ abstract class Element<T extends Widget> implements BuildContext {
 
   RenderObject findRenderObject() => renderObject;
 
-  Set<Type> _dependencies;
+  Map<Type, InheritedElement> _dependenciesByType;
+  Set<InheritedElement> _dependencies;
   InheritedWidget inheritFromWidgetOfExactType(Type targetType) {
-    if (_dependencies == null)
-      _dependencies = new Set<Type>();
-    _dependencies.add(targetType);
-    return ancestorWidgetOfExactType(targetType);
+    if (_dependenciesByType != null) {
+      InheritedElement cachedAncestor = _dependenciesByType[targetType];
+      if (cachedAncestor != null)
+        return cachedAncestor.widget;
+    }
+    Element ancestor = _parent;
+    while (ancestor != null && ancestor.widget.runtimeType != targetType)
+      ancestor = ancestor._parent;
+    if (ancestor != null) {
+      assert(ancestor is InheritedElement);
+      _dependenciesByType ??= <Type, InheritedElement>{};
```

Comment:
This isn't the best way of doing this.  In this approach, every widget has a separately allocated map that it manages.  Instead, a child should share the map with its parent because they'll usually be identical.  Inherited widgets should introduce new maps because they're the place where the maps will change.